### PR TITLE
Added limited optimizer support for Partitioned Hash Joins.

### PR DESCRIPTION
**Pull request by zuyu**
This PR allows repartition for non-stored, partitioned input relations for hash joins.

Repartition will generate the best partition attribute if both input relations have partitions that do not meet the join condition. Otherwise, repartition on one input relation will reuse the partition attribute from the other that meets the join condition.

### CODE REVIEW

**Comment by zuyu** [comment]
Assigned to @hbdeshmukh.

File: query_optimizer/ExecutionGenerator.hpp [hunk 1]
```diff
@@ -135,9 +135,11 @@ class ExecutionGenerator {
    */
   struct CatalogRelationInfo {
     CatalogRelationInfo(const QueryPlan::DAGNodeIndex producer_operator_index_in,
-                        const CatalogRelation *relation_in)
+                        const CatalogRelation *relation_in,
+                        const QueryContext::insert_destination_id output_destination_index_in)
```
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.hpp]: What do the ``_in`` suffix mean?

File: query_optimizer/tests/execution_generator/Partition.test [hunk 2]
```diff
@@ -15,19 +15,35 @@
 # specific language governing permissions and limitations
 # under the License.
 
-CREATE TABLE foo (id INT NULL,
-                  name CHAR(20))
+CREATE TABLE dim (id INT NULL,
```
Comment (hbdeshmukh) [inline query_optimizer/tests/execution_generator/Partition.test]: Optional, but it may be easier to read if the relation is named as ``dim_4_hash_partitions``.

File: query_optimizer/ExecutionGenerator.cpp [hunk 3]
```diff
@@ -679,13 +688,72 @@ void ExecutionGenerator::convertFilterJoin(const P::FilterJoinPtr &physical_plan
...
+                            probe_relation_info->output_destination_index));
 
   DCHECK(lip_filter_generator_ != nullptr);
   lip_filter_generator_->addFilterJoinInfo(physical_plan,
                                            build_filter_operator_index);
 }
 
+namespace {
+
+bool areSamePartitionSchemeHeaders(const PartitionSchemeHeader &lhs_partition_header,
```
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: These functions in the anonymous namespace may better be defined in some other source file. This cpp file is getting monstrously big, and any efforts to reduce its size should be welcome.

File: query_optimizer/ExecutionGenerator.cpp [hunk 4]
```diff
@@ -679,13 +688,72 @@ void ExecutionGenerator::convertFilterJoin(const P::FilterJoinPtr &physical_plan
...
+  switch (lhs_partition_header.getPartitionType()) {
+    case PartitionSchemeHeader::PartitionType::kHash:
+      return true;
+    case PartitionSchemeHeader::PartitionType::kRange: {
+      const vector<TypedValue> &lhs_ranges =
+          static_cast<const RangePartitionSchemeHeader&>(lhs_partition_header).getPartitionRangeBoundaries();
+      const vector<TypedValue> &rhs_ranges =
+          static_cast<const RangePartitionSchemeHeader&>(rhs_partition_header).getPartitionRangeBoundaries();
+
+      return lhs_ranges == rhs_ranges;
```
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: Can you help me understand how the comparison of two vector references will produce the desired result in this case?

File: query_optimizer/ExecutionGenerator.cpp [hunk 5]
```diff
@@ -679,13 +688,72 @@ void ExecutionGenerator::convertFilterJoin(const P::FilterJoinPtr &physical_plan
...
+
+  return false;
+}
+
+
+// Note that this method will be deprecated once the partition scheme header
+// supports multiple partition attributes.
+size_t chooseBestRepartitionAttributeIndex(const CatalogRelationStatistics &stats,
+                                           const vector<attribute_id> &join_attributes) {
+  size_t chose_attr_index = static_cast<size_t>(-1);
```
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: This would assign ``chose_attr_index`` the value of ``MAX_INT``. Why do we want that?

File: query_optimizer/ExecutionGenerator.cpp [hunk 6]
```diff
@@ -679,13 +688,72 @@ void ExecutionGenerator::convertFilterJoin(const P::FilterJoinPtr &physical_plan
...
+  for (std::size_t i = 0; i < join_attributes.size(); ++i) {
+    const attribute_id attr = join_attributes[i];
+    if (stats.hasNumDistinctValues(attr) &&
+        stats.getNumDistinctValues(attr) > chose_attr_num_distinct_values) {
+      chose_attr_index = i;
+      chose_attr_num_distinct_values = stats.getNumDistinctValues(attr);
+    }
+  }
+
+  return (chose_attr_index != static_cast<size_t>(-1)) ? chose_attr_index : 0;
```
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: I am unable to understand the point of this function. Why do we need to find the "best repartition attribute"? Assuming a join like ``R.a = S.b``, if ``R`` is partitioned on some other attribute ``c``, we would like to repartition it based on ``a``, so the only choice for an attribute on which we should perform the repartitioning is ``a``, isn't it? What's the decision got do do with the CatalogRelation stats and the number of distinct values of the attribute?

File: query_optimizer/ExecutionGenerator.hpp (hunk 1 above)
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.hpp]: It is used to name an argument in the constructor for a struct.

File: query_optimizer/ExecutionGenerator.cpp (hunk 4 above)
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: The comparison result checks every value in both vectors to ensure whether two have the same range partitions.

File: query_optimizer/ExecutionGenerator.cpp [hunk 7]
```diff
@@ -740,6 +805,202 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
...
+            }
+            break;
+          }
+        }
+        break;
+      case PartitionSchemeHeader::PartitionType::kHash: {
+        const attribute_id build_partition_attr = build_partition_scheme_header.getPartitionAttributeId();
+        if (find(build_attribute_ids.begin(), build_attribute_ids.end(), build_partition_attr) !=
+                build_attribute_ids.end()) {
+          // BuildRelation has an useful partition.
```
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: Grammar: an -> a

File: query_optimizer/ExecutionGenerator.cpp [hunk 8]
```diff
@@ -740,6 +805,202 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
...
+
+    switch (build_partition_scheme_header.getPartitionType()) {
+      case PartitionSchemeHeader::PartitionType::kRange:
+        build_needs_repartition = true;
+
+        switch (probe_partition_scheme_header.getPartitionType()) {
+          case PartitionSchemeHeader::PartitionType::kRange:
+            probe_needs_repartition = true;
+            break;
+          case PartitionSchemeHeader::PartitionType::kHash: {
```
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: Should be good to add a default case here.

File: query_optimizer/ExecutionGenerator.cpp [hunk 9]
```diff
@@ -740,6 +805,202 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
...
+            const attribute_id probe_partition_attr = probe_partition_scheme_header.getPartitionAttributeId();
+            if (find(probe_attribute_ids.begin(), probe_attribute_ids.end(), probe_partition_attr) !=
+                    probe_attribute_ids.end()) {
+              needs_swap = true;
```
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: Why do we need to swap here? There's no cardinality check performed here unlike line 863.

File: query_optimizer/ExecutionGenerator.cpp [hunk 10]
```diff
@@ -740,6 +805,202 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
...
+      case PartitionSchemeHeader::PartitionType::kHash: {
+        const attribute_id build_partition_attr = build_partition_scheme_header.getPartitionAttributeId();
+        if (find(build_attribute_ids.begin(), build_attribute_ids.end(), build_partition_attr) !=
+                build_attribute_ids.end()) {
+          // BuildRelation has an useful partition.
+          switch (probe_partition_scheme_header.getPartitionType()) {
+            case PartitionSchemeHeader::PartitionType::kRange:
+              probe_needs_repartition = true;
+              break;
+            case PartitionSchemeHeader::PartitionType::kHash: {
```
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: Again, a default case should make the code more extensible.

Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: I intend not to add a default case, so that once a new partition scheme is added, we will have compile errors to force to revisit how the new partition scheme fits in this method.

File: query_optimizer/ExecutionGenerator.cpp (hunk 9 above)
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: In a hash join, if the build relation uses range partition, while the probe uses hash partition, we swap two relations, and then repartition the probe (aka, the original build) relation.

File: query_optimizer/ExecutionGenerator.cpp (hunk 6 above)
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: In a hash join, if both relations are using range partitions, we need to repartition both relations. Here we use this method to pick up the best single attribute for the new hash repartition. By best, we mean the maximum number of distinct values to minimize the hash collision.

Finally, once we support multiple partition attributes, we don't need this method.

File: query_optimizer/ExecutionGenerator.cpp (hunk 5 above)
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: That's just an initial value.

File: query_optimizer/ExecutionGenerator.cpp (hunk 3 above)
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Unless these functions are useful to other files, we keep them in the same file as the caller.

Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: I had a discussion with @jianqiao on this issue. He thinks that the partitioning related code should be moved to different functions at least. That way the overall code is easier to read and is also extensible. He's fine with the anonymous namespace arrangement. 

An example is all the new code in ``convertHashJoin`` can be moved to separate function(s) and then called from ``convertHashJoin``.

File: query_optimizer/ExecutionGenerator.cpp (hunk 10 above)
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: Are you sure that the compiler will throw an error saying that we didn't handle all the cases in the switch statement? I would recommend a safe approach where we add a default case that throws a fatal error.

File: query_optimizer/ExecutionGenerator.cpp (hunk 9 above)
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: I am not sure if I agree. Consider the following counterexample. 
Original build table - cardinality 10k, range partitioned. 
Original probe table - cardinality 1M, hash partitioned. 

By the above logic, you will swap the build and probe relations, thereby building a giant hash table for 1M entries. You are going to repartition the new probe relation anyway, so that cost isn't avoided.

File: query_optimizer/ExecutionGenerator.cpp [hunk 11]
```diff
@@ -740,6 +805,202 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
...
+              } else {
+                probe_needs_repartition = true;
+              }
+              break;
+            }
+          }
+        }
+        break;
+      }
+    }
```
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: I would highly recommend two things for the switch statement:
1. Add several comments in between to describe what's happening. I felt lost while reading it.
2. Moving the switch statement to a separate method.

File: query_optimizer/ExecutionGenerator.cpp [hunk 12]
```diff
@@ -740,6 +805,202 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
...
+      S::InsertDestination *build_insert_destination_proto =
+          query_context_proto_->mutable_insert_destinations(build_relation_info->output_destination_index);
+
+      build_insert_destination_proto->set_insert_destination_type(S::InsertDestinationType::PARTITION_AWARE);
+      build_insert_destination_proto->MutableExtension(S::PartitionAwareInsertDestination::partition_scheme)
+        ->MergeFrom(build_repartition_scheme->getProto());
+
+      CatalogRelation *mutable_build_relation =
+          catalog_database_->getRelationByIdMutable(build_relation->getID());
+      mutable_build_relation->setPartitionScheme(build_repartition_scheme.release());
```
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: May be I missed some of your older PRs, but can you explain how we will physically repartition the relations?

**Comment by hbdeshmukh** [comment]
One case which is currently not considered is the following:

Imagine a join between two relations. The probe side is a result of a selection on a base relation. The probe's base relation is hash partitioned on attribute ``a``. There are two cases:
1. In the projection, the attribute ``a`` is left out, i.e. we don't project it. 
2. We project the attribute ``a``. 

If you look at the code of ``convertSelection``, the output of the selection has no partitioning scheme. In case two, we can carry forward the partitioning scheme from the base table to the output of the selection. That way, you can avoid the repartitioning step for the probe side.

To summarize, you need to add a check in ``convertSelection`` that can identify case 1 and case 2 mentioned earlier.

File: query_optimizer/ExecutionGenerator.cpp (hunk 9 above)
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: I get your point, but the partition number is set by us. So we won't have such insane number, but I could add a TODO here.

File: query_optimizer/ExecutionGenerator.cpp (hunk 10 above)
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Yes, it is `-Wswitch` and then becomes an error for us.

File: query_optimizer/ExecutionGenerator.cpp (hunk 3 above)
Comment (pateljm) [inline query_optimizer/ExecutionGenerator.cpp]: +1 for the suggestion to move the partition code to a different file. It is (related) but separate module.

File: query_optimizer/ExecutionGenerator.cpp (hunk 12 above)
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: We modify the existing `SelectOperator` to have a new `PartitionAwareInsertDestination` for repartition, if the input relation is not a base table. Otherwise, we will create one, but not part of this PR.

File: query_optimizer/ExecutionGenerator.cpp (hunk 9 above)
Comment (hbdeshmukh) [inline query_optimizer/ExecutionGenerator.cpp]: It's clearly not an insane number. I said **cardinality** and not **number of partitions**. SSB Lineorder table for SF100 has a cardinality of 600M, so 1M is not insane by any means.

Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Now I got your point. I'll do another swap if needed, so that the build side has a less cardinality. Thanks!

**Comment by zuyu** [comment]
Close it temporally.
